File: golang/direct_e2ee/ratchet.go
package directe2ee

import (
	"crypto/sha256"

	"golang.org/x/crypto/chacha20poly1305"
)

const MaxSkip uint32 = 1000

// ChainStep contains derived message material for a single ratchet step.
type ChainStep struct {
	MessageKey   [32]byte
	Nonce        [12]byte
	NextChainKey [32]byte
}

// DeriveChainStep derives message key, nonce, and next chain key.
func DeriveChainStep(chainKey [32]byte) ChainStep {
	nextChainKey := digestWithLabel(chainKey, []byte("ANP Direct E2EE v1 Next Chain Key"))
	messageKey := digestWithLabel(chainKey, []byte("ANP Direct E2EE v1 Message Key"))
	nonceMaterial := digestWithLabel(chainKey, []byte("ANP Direct E2EE v1 Message Nonce"))
	var nonce [12]byte
	copy(nonce[:], nonceMaterial[:12])
	return ChainStep{MessageKey: messageKey, Nonce: nonce, NextChainKey: nextChainKey}
}

// EncryptWithStep encrypts plaintext with the derived step.
func EncryptWithStep(step ChainStep, plaintext []byte, aad []byte) ([]byte, error) {
	aead, err := chacha20poly1305.New(step.MessageKey[:])
	if err != nil {
		return nil, cryptoError("invalid ChaCha20-Poly1305 key")
	}
	return aead.Seal(nil, step.Nonce[:], plaintext, aad), nil
}

// DecryptWithStep decrypts ciphertext with the derived step.
func DecryptWithStep(step ChainStep, ciphertext []byte, aad []byte) ([]byte, error) {
	aead, err := chacha20poly1305.New(step.MessageKey[:])
	if err != nil {
		return nil, cryptoError("invalid ChaCha20-Poly1305 key")
	}
	plaintext, err := aead.Open(nil, step.Nonce[:], ciphertext, aad)
	if err != nil {
		return nil, cryptoError("failed to decrypt ciphertext")
	}
	return plaintext, nil
}

func digestWithLabel(chainKey [32]byte, label []byte) [32]byte {
	hash := sha256.Sum256(append(append([]byte(nil), label...), chainKey[:]...))
	return hash
}

File: rust/src/direct_e2ee/ratchet.rs
use super::errors::DirectE2eeError;
use ring::aead::{Aad, LessSafeKey, Nonce, UnboundKey, CHACHA20_POLY1305};
use sha2::{Digest, Sha256};

pub const MAX_SKIP: u32 = 1000;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ChainStep {
    pub message_key: [u8; 32],
    pub nonce: [u8; 12],
    pub next_chain_key: [u8; 32],
}

pub fn derive_chain_step(chain_key: &[u8; 32]) -> ChainStep {
    let next_chain_key = digest_with_label(chain_key, b"ANP Direct E2EE v1 Next Chain Key");
    let message_key = digest_with_label(chain_key, b"ANP Direct E2EE v1 Message Key");
    let nonce_material = digest_with_label(chain_key, b"ANP Direct E2EE v1 Message Nonce");
    let mut nonce = [0u8; 12];
    nonce.copy_from_slice(&nonce_material[..12]);
    ChainStep {
        message_key,
        nonce,
        next_chain_key,
    }
}

pub fn encrypt_with_step(
    step: &ChainStep,
    plaintext: &[u8],
    aad: &[u8],
) -> Result<Vec<u8>, DirectE2eeError> {
    let unbound = UnboundKey::new(&CHACHA20_POLY1305, &step.message_key)
        .map_err(|_| DirectE2eeError::crypto("invalid ChaCha20-Poly1305 key"))?;
    let key = LessSafeKey::new(unbound);
    let nonce = Nonce::assume_unique_for_key(step.nonce);
    let mut buffer = plaintext.to_vec();
    key.seal_in_place_append_tag(nonce, Aad::from(aad), &mut buffer)
        .map_err(|_| DirectE2eeError::crypto("failed to encrypt ciphertext"))?;
    Ok(buffer)
}

pub fn decrypt_with_step(
    step: &ChainStep,
    ciphertext: &[u8],
    aad: &[u8],
) -> Result<Vec<u8>, DirectE2eeError> {
    let unbound = UnboundKey::new(&CHACHA20_POLY1305, &step.message_key)
        .map_err(|_| DirectE2eeError::crypto("invalid ChaCha20-Poly1305 key"))?;
    let key = LessSafeKey::new(unbound);
    let nonce = Nonce::assume_unique_for_key(step.nonce);
    let mut buffer = ciphertext.to_vec();
    let plaintext = key
        .open_in_place(nonce, Aad::from(aad), &mut buffer)
        .map_err(|_| DirectE2eeError::crypto("failed to decrypt ciphertext"))?;
    Ok(plaintext.to_vec())
}

fn digest_with_label(chain_key: &[u8; 32], label: &[u8]) -> [u8; 32] {
    let mut hasher = Sha256::new();
    hasher.update(label);
    hasher.update(chain_key);
    hasher.finalize().into()
}

#[cfg(test)]
mod tests {
    use super::{decrypt_with_step, derive_chain_step, encrypt_with_step};

    #[test]
    fn chain_step_encrypts_and_decrypts() {
        let step = derive_chain_step(&[9u8; 32]);
        let ciphertext = encrypt_with_step(&step, b"hello", br#"{"aad":true}"#)
            .expect("encrypt");
        let plaintext = decrypt_with_step(&step, &ciphertext, br#"{"aad":true}"#)
            .expect("decrypt");
        assert_eq!(plaintext, b"hello");
    }
}
